File: LeetCode/Java/src/problems/p0226_InvertBinaryTree/P0226.java
package problems.p0226_InvertBinaryTree;

import auxiliary.TreeNode;

/*
 * Given the root of a binary tree, invert the tree, and return its root.
 *
 * Example 1:
 * Input: root = [4,2,7,1,3,6,9]
 * Output: [4,7,2,9,6,3,1]
 *
 * Example 2:
 * Input: root = [2,1,3]
 * Output: [2,3,1]
 *
 * Example 3:
 * Input: root = []
 * Output: []
 *
 * Constraints:
 * The number of nodes in the tree is in the range [0, 100].
 * -100 <= Node.val <= 100
 */

public class P0226 {
    public TreeNode invertTree(TreeNode root) {
        if (root == null) {
            return null;
        }
        /*
         * Solution v1: Iterative
         *
         * Runtime: 0 ms @ (beats) 100%
         * Memory Usage: 36.2 MB @ (beats) 96.72%
         */
        // Queue<TreeNode> queue = new LinkedList<>();
        // queue.add(root);
        // while (!queue.isEmpty()) {
        //     TreeNode node = queue.poll();
        //     TreeNode temp = node.left;
        //     node.left = node.right;
        //     node.right = temp;
        //     if (node.left != null) {
        //         queue.add(node.left);
        //     }
        //     if (node.right != null) {
        //         queue.add(node.right);
        //     }
        // }
        // return root;

        /*
         * Solution v2: Recursive
         *
         * Runtime: 0 ms @ (beats) 100%
         * Memory Usage: 36.6 MB @ (beats) 54.60%
         */
        TreeNode temp = root.left;
        root.left = root.right;
        root.right = temp;
        if (root.left != null) {
            invertTree(root.left);
        }
        if (root.right != null) {
            invertTree(root.right);
        }
        return root;
    }

}

File: LeetCode/Python/solutions/p0226.py
import collections
from typing import Optional

from Python.auxiliary.TreeNode import TreeNode


class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        #
        # Solution v1: Recursive
        #
        # Runtime: 43 ms @ (beats) 15.93%
        # Memory Usage: 13.7 MB @ (beats) 94.98%
        #
        # ans = None
        #
        # if not root:
        #     return ans
        #
        # ans = TreeNode()
        # ans.val = root.val
        #
        # ans.left = self.invertTree(root.right)
        # ans.right = self.invertTree(root.left)
        #
        # return ans

        #
        # Solution v1.1: Recursive in place
        #
        # Runtime: 34 ms @ (beats) 59.87%
        # Memory Usage: 13.9 MB @ (beats) 7.55%
        #

        # if not root:
        #     return None
        #
        # temp = root.left
        #
        # root.left = root.right
        # root.right = temp
        #
        # if root.left:
        #     self.invertTree(root.left)
        #
        # if root.right:
        #     self.invertTree(root.right)
        #
        # return root

        #
        # Solution v2: Iterative
        #
        # Runtime: 35 ms @ (beats) 52.26%
        # Memory Usage: 14 MB @ (beats) 7.55%
        #

        if not root:
            return None

        q = collections.deque()
        q.append(root)

        while q:
            node = q.popleft()

            if not node:
                continue

            temp = node.left

            node.left = node.right
            node.right = temp

            q.append(node.left)
            q.append(node.right)

        return root